### PR TITLE
Add an option to force using hostname instead of FQDN in the resource detection processor

### PR DESCRIPTION
**Pull request by hprajapati-splunk**
**Description:** 
This PR adds an option to force using hostname instead of FQDN in the resource detection processor.

Users can set `hostname_sources` to `["os"]` to force the processor to use 'hostname'.

This is an opt-in functionality so there are no breaking changes to existing user configurations   

**Testing:** 
Existing unit tests have been updated and new unit tests have been added to cover new changes.

**Documentation:**
README has been updated to show how to use the new custom 'system' configurations

### CODE REVIEW

File: processor/resourcedetectionprocessor/README.md [hunk 1]
```diff
@@ -17,6 +17,14 @@ details of which are currently pending confirmation in the OpenTelemetry specifi
     * host.name
     * os.type
 
+system detector custom configuration example:
+```yaml
+detectors: ["system"]
+system:
+    # Force the detector to use 'hostname' instead of FQDN
+    use_hostname: true
```
Comment (mx-psi) [inline processor/resourcedetectionprocessor/README.md:25]: What about something like this?
```suggestion
    system_source: hostname
```

That way if we add further options to the system detector the new option will still be useful. And we can have an `auto` default for the current behavior (try FQDN and fall back to hostname). Does that make sense to you? (I am not super convinced myself so please push back if this has some flaw 😄)

File: processor/resourcedetectionprocessor/README.md [hunk 2]
```diff
@@ -17,6 +17,14 @@ details of which are currently pending confirmation in the OpenTelemetry specifi
     * host.name
     * os.type
 
+system detector custom configuration example:
```
Comment (mx-psi) [inline processor/resourcedetectionprocessor/README.md:20]: This should have the same indentation as the `* System metadata` bullet point above. Also
```suggestion
System detector custom configuration example:
```

**Comment by mx-psi** [comment]
Also @hprajapati-splunk you need to sign the CLA so that we can accept your contribution, PTAL at the bot message above!

File: processor/resourcedetectionprocessor/README.md (hunk 1 above)
Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/README.md]: Yes, that makes sense. I have added the suggested changes.

File: processor/resourcedetectionprocessor/README.md (hunk 2 above)
Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/README.md]: I have added the suggested changes.

File: processor/resourcedetectionprocessor/README.md (hunk 1 above)
Comment (dmitryax) [reply processor/resourcedetectionprocessor/README.md]: I don't think that `system_source` is a proper name. It's a source for fetching `host.name` attribute value. 

Can we maybe call it `hostname_sources` and have the value as a list of sources in order of fetching priority. The default will be ["fqdn", "hostname"]. For the use case when FQDN is not needed we set it to `hostname_sources: ["hostname"]`. 

@mx-psi @hprajapati-splunk what do you think?

Comment (dmitryax) [reply processor/resourcedetectionprocessor/README.md]: We can consider other names for the sources values like `hostname_sources: ["dns", "os"]` for the default. Not sure tho, happy to discuss other options

Comment (mx-psi) [reply processor/resourcedetectionprocessor/README.md]: `hostname_sources` sounds better to me than `system_source`and passing a priority list makes sense, they both seem like good suggestions

Comment (dmitryax) [reply processor/resourcedetectionprocessor/README.md]: ok, sounds good. if everyone agrees, @hprajapati-splunk let's use the config with `hostname_sources: ["dns", "os"]` by default

Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/README.md]: @dmitryax @mx-psi 
If we use a priority list then few questions arise like:
    1. What if any value in the given list is invalid like: `["dns", "invalid"]` ?
    2. If all the values in the priority list failed for ex. `["dns", "os"]` is the passed list but both methods failed to fetch the 
        required value then shall we return one error message or all the errors that we got from all the methods(errors from 
        methods which fetch `FQDN` and `hostname`?

Comment (mx-psi) [reply processor/resourcedetectionprocessor/README.md]: >  1. What if any value in the given list is invalid like: ["dns", "invalid"] ?

I think the most sensible way is to check this at configuration validation time and return an error if there are any invalid entries.

> 2. If all the values in the priority list failed for ex. ["dns", "os"] is the passed list but both methods failed to fetch the required value then shall we return one error message or all the errors that we got from all the methods(errors from methods which fetch FQDN and hostname?

Yes, you can use `consumererror.Combine` for this.

Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/README.md]: @dmitryax @mx-psi 
I have added the suggested changes. Please, review the same.

File: processor/resourcedetectionprocessor/internal/system/system.go [hunk 3]
```diff
@@ -31,21 +32,35 @@ const (
...
 
 // NewDetector creates a new system metadata detector
-func NewDetector(p component.ProcessorCreateSettings, _ internal.DetectorConfig) (internal.Detector, error) {
-	return &Detector{provider: &systemMetadataImpl{}, logger: p.Logger}, nil
+func NewDetector(p component.ProcessorCreateSettings, dcfg internal.DetectorConfig) (internal.Detector, error) {
+	cfg := dcfg.(Config)
+	err := cfg.Validate()
+	if err != nil {
+		return nil, err
+	}
```
Comment (mx-psi) [inline processor/resourcedetectionprocessor/internal/system/system.go:55]: This should be called on a `Validate` method of the top-level configuration (there is no method as of now, it should have this type signature https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/10b008641faa7300222ff608a6337a6a8398caa0/receiver/jaegerreceiver/config.go#L87 ). That way this can be validated without creating the detector (e.g. in a dry run).

File: processor/resourcedetectionprocessor/internal/system/system.go [hunk 4]
```diff
@@ -54,18 +69,36 @@ func (d *Detector) Detect(_ context.Context) (resource pdata.Resource, schemaURL
...
+		getHostname := hostnameSourcesMap[source]
+		hostname, err = getHostname(d)
+		if err == nil {
+			attrs.InsertString(conventions.AttributeHostName, hostname)
+			attrs.InsertString(conventions.AttributeOSType, osType)
+
+			return res, conventions.SchemaURL, nil
+		} else {
+			errs = append(errs, err)
 		}
```
Comment (mx-psi) [inline processor/resourcedetectionprocessor/internal/system/system.go:82]: Unnecessary else
```suggestion
		}
		errs = append(errs, err)
```

File: processor/resourcedetectionprocessor/internal/system/config.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+package system
+
+import "fmt"
+
+// Config defines user-specified configurations unique to the system detector
+type Config struct {
+	// The HostnameSources is a priority list of sources from which hostname will be fetched.
+	// In case of the error in fetching hostname from source,
+	// the next source from the list will be considered.(**default**: `["dns", "os"]`)
```
Comment (mx-psi) [inline processor/resourcedetectionprocessor/internal/system/config.go:23]: We should specify somewhere the valid values to pass to this

**Review by mx-psi — COMMENTED** [review]
Thanks for bearing with us on the review :) Behavior-wise this looks good to me but I left a couple of minor comments about the code. Also, note that CI will be broken until #5076 gets merged (sorry about that!)

File: processor/resourcedetectionprocessor/internal/system/system.go (hunk 3 above)
Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/internal/system/system.go]: I have added the suggested changes. Please, review the same.

File: processor/resourcedetectionprocessor/internal/system/system.go (hunk 4 above)
Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/internal/system/system.go]: I have added the suggested changes. Please, review the same.

File: processor/resourcedetectionprocessor/internal/system/config.go (hunk 5 above)
Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/internal/system/config.go:23]: I have added the suggested changes. Please, review the same.

**Review by mx-psi — APPROVED** [review]
Looks good, thanks again! I would like @dmitryax to have a look at it too

File: processor/resourcedetectionprocessor/factory.go [hunk 6]
```diff
@@ -158,6 +158,10 @@ func (f *factory) getResourceDetectionProcessor(
 	cfg config.Processor,
 ) (*resourceDetectionProcessor, error) {
 	oCfg := cfg.(*Config)
+	err := oCfg.Validate()
+	if err != nil {
+		return nil, err
+	}
```
Comment (dmitryax) [inline processor/resourcedetectionprocessor/factory.go:164]: This part is not needed AFAIK. `cfg.Validate()` is called from the core config validation. But still please verify it by trying this out.

File: processor/resourcedetectionprocessor/internal/system/system.go [hunk 7]
```diff
@@ -31,21 +32,34 @@ const (
...
 }
 
 // NewDetector creates a new system metadata detector
-func NewDetector(p component.ProcessorCreateSettings, _ internal.DetectorConfig) (internal.Detector, error) {
-	return &Detector{provider: &systemMetadataImpl{}, logger: p.Logger}, nil
+func NewDetector(p component.ProcessorCreateSettings, dcfg internal.DetectorConfig) (internal.Detector, error) {
+	cfg := dcfg.(Config)
+	if len(cfg.HostnameSources) == 0 {
+		cfg.HostnameSources = []string{"dns", "os"}
+	}
```
Comment (dmitryax) [inline processor/resourcedetectionprocessor/internal/system/system.go:54]: Instead of patching the config here, can we add `system -> hostname_sources -> ["dns", "os"']` in `createDefaultConfig` even if `system` detector is not enabled by default?

Then we can also reject `hostname_sources: []` config during the validation step

File: processor/resourcedetectionprocessor/factory_test.go [hunk 8]
```diff
@@ -51,7 +52,12 @@ func TestInvalidConfig(t *testing.T) {
 	factory := NewFactory()
 	cfg := factory.CreateDefaultConfig()
 	oCfg := cfg.(*Config)
-	oCfg.Detectors = []string{"not-existing"}
+	oCfg.Detectors = []string{"not-existing", "system"}
+	oCfg.DetectorConfig = DetectorConfig{
+		SystemConfig: system.Config{
+			HostnameSources: []string{"dns", "invalid"},
+		},
+	}
```
[the remark this answers is not in the document]
Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/factory_test.go]: Created another test case to test detector invalid `detectorConfig` separately.

File: processor/resourcedetectionprocessor/config_test.go [hunk 9]
```diff
@@ -93,6 +107,18 @@ func TestGetConfigFromType(t *testing.T) {
...
+		{
+			name:         "Get System Config",
+			detectorType: system.TypeStr,
+			inputDetectorConfig: DetectorConfig{
+				SystemConfig: system.Config{
+					HostnameSources: []string{"dns", "os"},
+				},
+			},
+			expectedConfig: system.Config{
+				HostnameSources: []string{"dns", "os"},
```
[the remark this answers is not in the document]
Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/config_test.go]: Changed `HostnameSources` to `os`.

File: processor/resourcedetectionprocessor/internal/system/system.go [hunk 10]
```diff
@@ -54,18 +68,35 @@ func (d *Detector) Detect(_ context.Context) (resource pdata.Resource, schemaURL
...
+	for _, source := range d.hostnameSources {
+		getHostFromSource := hostnameSourcesMap[source]
+		hostname, err = getHostFromSource(d)
+		if err == nil {
+			attrs.InsertString(conventions.AttributeHostName, hostname)
+			attrs.InsertString(conventions.AttributeOSType, osType)
+
+			return res, conventions.SchemaURL, nil
 		}
+		errs = append(errs, err)
```
[the remark this answers is not in the document]
Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/internal/system/system.go]: I have added the suggested changes. Please, review the same.

File: processor/resourcedetectionprocessor/internal/system/system.go [hunk 11]
```diff
@@ -54,18 +68,35 @@ func (d *Detector) Detect(_ context.Context) (resource pdata.Resource, schemaURL
...
+// getHostname returns OS hostname
+func getHostname(d *Detector) (string, error) {
+	hostname, err := d.provider.Hostname()
+	if err != nil {
+		return "", fmt.Errorf("failed getting OS hostname: %w", err)
+	}
+	return hostname, nil
+}
+
+// getFQDN returns FQDN
```
[the remark this answers is not in the document]
Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/internal/system/system.go]: I have added the suggested changes. Please, review the same.

File: processor/resourcedetectionprocessor/factory.go (hunk 6 above)
Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/factory.go]: Yes, It is not needed and I have verified the same.

File: processor/resourcedetectionprocessor/internal/system/system.go (hunk 7 above)
Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/internal/system/system.go:54]: It will be done once https://github.com/open-telemetry/opentelemetry-collector/issues/4001 gets resolved. So, for now added `TODO` for the same.

File: processor/resourcedetectionprocessor/README.md [hunk 12]
```diff
@@ -17,6 +17,18 @@ details of which are currently pending confirmation in the OpenTelemetry specifi
     * host.name
     * os.type
 
+* System detector custom configuration example:
```
Comment (dmitryax) [inline processor/resourcedetectionprocessor/README.md:20]:
```suggestion
By default `host.name` is being set to FQND if possible, and a hostname provided by OS used as fallback.
This logic can be changed with `hostname_sources` configuration which is set to `["dns", "os"]` by default.

Use the following config in order to use hostname provided by OS only: 
```

File: processor/resourcedetectionprocessor/README.md [hunk 13]
```diff
@@ -17,6 +17,18 @@ details of which are currently pending confirmation in the OpenTelemetry specifi
...
+    ```yaml
+    detectors: ["system"]
+    system:
+        # A priority list of sources to get the hostname(**default**: `["dns", "os"]`)
+        hostname_sources: ["dns", "os"]
```
Comment (dmitryax) [inline processor/resourcedetectionprocessor/README.md:25]:
```suggestion
        hostname_sources: ["os"]
```

File: processor/resourcedetectionprocessor/factory_test.go [hunk 14]
```diff
@@ -51,7 +52,12 @@ func TestInvalidConfig(t *testing.T) {
 	factory := NewFactory()
 	cfg := factory.CreateDefaultConfig()
 	oCfg := cfg.(*Config)
-	oCfg.Detectors = []string{"not-existing"}
+	oCfg.Detectors = []string{"not-existing", "system"}
+	oCfg.DetectorConfig = DetectorConfig{
+		SystemConfig: system.Config{
+			HostnameSources: []string{"dns"},
+		},
+	}
```
Comment (dmitryax) [inline processor/resourcedetectionprocessor/factory_test.go:60]: This looks like unneeded change. Can we discard it?

Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/factory_test.go:60]: You mean discard `SystemConfig` ?

File: processor/resourcedetectionprocessor/README.md (hunk 12 above)
Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/README.md]: I have added the suggested changes. Please, review the same.

File: processor/resourcedetectionprocessor/README.md (hunk 13 above)
Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/README.md]: I have added the suggested changes. Please, review the same.

File: processor/resourcedetectionprocessor/factory_test.go (hunk 14 above)
Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/factory_test.go]: Yes, I have discarded this changes.

**Comment by bogdandrutu** [comment]
@dmitryax PTAL also set the "ready-to-merge" label when you are done :)

File: processor/resourcedetectionprocessor/README.md [hunk 15]
```diff
@@ -17,6 +17,22 @@ details of which are currently pending confirmation in the OpenTelemetry specifi
     * host.name
     * os.type
 
+* By default `host.name` is being set to FQND if possible, and a hostname provided by OS used as fallback.
```
Comment (dmitryax) [inline processor/resourcedetectionprocessor/README.md:20]: Please remove the list bullet since it's related to the current list section (not a new one) :)

File: processor/resourcedetectionprocessor/README.md [hunk 16]
```diff
@@ -17,6 +17,22 @@ details of which are currently pending confirmation in the OpenTelemetry specifi
...
+This logic can be changed with `hostname_sources` configuration which is set to `["dns", "os"]` by default.
+
+Use the following config to avoid getting FQDN and apply hostname provided by OS only:
+
+    ```yaml
+    detectors: ["system"]
+    system:
+        # A priority list of sources to get the hostname(**default**: `["dns", "os"]`)
```
Comment (dmitryax) [inline processor/resourcedetectionprocessor/README.md:28]: I don't think we need this line anymore

File: processor/resourcedetectionprocessor/factory.go [hunk 17]
```diff
@@ -158,7 +160,6 @@ func (f *factory) getResourceDetectionProcessor(
 	cfg config.Processor,
 ) (*resourceDetectionProcessor, error) {
 	oCfg := cfg.(*Config)
-
```
Comment (dmitryax) [inline processor/resourcedetectionprocessor/factory.go:161]: nit: unrelated change

File: processor/resourcedetectionprocessor/README.md (hunk 15 above)
Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/README.md]: I have removed the list bullet.

File: processor/resourcedetectionprocessor/README.md (hunk 16 above)
Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/README.md]: I have removed that line.

File: processor/resourcedetectionprocessor/factory.go (hunk 17 above)
Comment (hprajapati-splunk) [reply processor/resourcedetectionprocessor/factory.go]: I have discarded this change.

**Review by dmitryax — APPROVED** [review]

File: processor/resourcedetectionprocessor/README.md [hunk 18]
```diff
@@ -17,6 +17,21 @@ details of which are currently pending confirmation in the OpenTelemetry specifi
     * host.name
     * os.type
 
+By default `host.name` is being set to FQND if possible, and a hostname provided by OS used as fallback.
```
Comment (tigrannajaryan) [inline processor/resourcedetectionprocessor/README.md:20]:
```suggestion
By default `host.name` is being set to FQDN if possible, and a hostname provided by OS used as fallback.
```
